### ISSUE TITLE
HighDPI: different scaling of Graphics in WinForms and Direct2D engines.

### ISSUE THREAD

**Issue by e673** [picoe/Eto#1141]
I'm developing a DPI-aware application which has a `Drawable` element with custom `OnPaint` implementation.

In `OnPaint` method I apply scaling transform to `e.Graphics` in order to get one-by-one correspondence between points and physical pixels. I get the ratio of scaling transform from `Screen.LogicalPixelSize` property.

Everything works fine with Wpf, WinForms, Gtk engines but I've got a problem with Direct2D engine.
When using Direct2D engine, I receive scaled `e.Graphics` object and have to calculate the exact scale ratio as a ratio between control size and clipping rectangle size.

Consider a control with a width of 2390. In `OnPaint` method I receive `e.Graphics` instance and apply scaling transform corresponding to actual scaling factor. I expect a clipping rectangle `(0, 0, 2390, *)` but got `(0,0, 2388, *)`.

I've analyzed `Eto.Direct2D.GraphicsHandler.ClipBounds` property. The problem seems to be in rounding and errorneous clip rectange calculation in `Eto.Direct2D.GraphicsHandler`.

1. The values are floating point but they are rounded in the middle of calculations in `ToEto()` method.
Changing `ToEto` implementation to the following
```cs
public static PointF ToEto(this s.Vector2 value)
{
    return new PointF(value.X, value.Y);
}
```
solves the rounding problem.

2. Calculation of new rectangle size using `EndLocation` is also wrong:
`(Width - 1) * scale + 1 != Width * scale`.

The correct implementation of `ClipBounds` property should be:

```cs
public RectangleF ClipBounds
{
	get
	{
		if (Control != null)
		{
			// not very efficient, but works
			s.Matrix3x2 transform = currentTransform; // Control.Transform; - the result should not depend on PixelOffsetMode
			transform.Invert();
			var start = s.Matrix3x2.TransformPoint(transform, clipBounds.Location.ToDx()).ToEto();
			var end = s.Matrix3x2.TransformPoint(transform, clipBounds.BottomRight.ToDx()).ToEto();
			return new RectangleF(start, new SizeF(end.X - start.X, end.Y - start.Y));
		}
		else
			return new RectangleF();
	}
}
```

**Comment by e673** [comment]
I was also expected `PixelOffsetMode` to be applied after all the transformations, not before. But considering issues #349, #350 seems it's not a bug but a feature. Changing its logic would break a lot of code. So to render graphics with `LogicalPixelSize` other than 1, it is better to use `PixelOffsetMode.Half` and draw the lines manually with adding `0.5f`.

**Comment by cwensley** [comment]
Hey @e673, 

Thanks for submitting the issue (and the subsequent fix for some of the issues).  As for the PixelOffsetMode, I think you're correct in that it should be applied after all transformations.  I think there may still be bugs still in its implementation.

If we can narrow down a small test that demonstrates expected behaviour vs. what is happening we can hopefully get that fixed.

**Comment by cwensley** [comment]
Fixed with #1143
